### ISSUE TITLE
[Performance]: guided generation is very slow in offline mode

### ISSUE THREAD

**Issue by stas00** [vllm-project/vllm#8313]
### Proposal to improve performance

With a single request / online mode I'm getting:

- no guided 300 tok/sec
- `outlines` 150 tok/sec (2x slower)
- `lm-format-enforcer` 90 tok/sec (~3x slower)

with offline mode I get:
- `outlines` **is about 10-20x slower than no guided generation**
- `lm-format-enforcer` is about 4x faster than `outlines` (note that it is slower than `outlines` for online)

for online I was using this schema:

```
json_template = {
    "type": "object",
    "properties": {
        "criteria": {"type": "array", "items": {"type": "string"}, "minItems": 1},
        "response": { "type": "string" }
    },
    "required": ["criteria", "response"]
}
```

for offline I was using an even simpler schema:
```

{
   "type":"object",
   "properties":{
      "name":{
         "type":"string", "minLength":2, "maxLength":5
      },
      "age":{
         "type":"integer"
      }
   },
   "required":[ "name", "age"]
}
```
the huge performance hit in the offline mode is very strange for both backends.

2x slow down in the online mode is pretty bad too as it's already a huge impact. The offline mode can actually tolerate 2x no problem as there is no human in the loop, but 10-20x is a way impractical.

`vllm=0.6.0` and `outlines==0.0.46`

**Comment by Quang-elec44** [comment]
@stas00 In my experience, guided generation is always slower than normal. I recommend you try `sglang` instead. Sglang achieves better throughput than vLLM, but the guided generation is still slower.

**Comment by stas00** [comment]
Thank you for this suggestion, @Quang-elec44 - I understand that it'll be slower, but it should be marginally slower, not 20x slower. Possibly some problem in the integration?

**Comment by robertgshaw2-neuralmagic** [comment]
> Thank you for this suggestion, @Quang-elec44 - I understand that it'll be slower, but it should be marginally slower, not 20x slower. Possibly some problem in the integration?

@stas00 is this a new issue on v0.6.0?

**Comment by robertgshaw2-neuralmagic** [comment]
@stas00 - looks like someone has just fixed the issue:
- https://github.com/vllm-project/vllm/pull/8308

**Comment by robertgshaw2-neuralmagic** [comment]
> @stas00 In my experience, guided generation is always slower than normal. I recommend you try `sglang` instead. Sglang achieves better throughput than vLLM, but the guided generation is still slower.

Hey @Quang-elec44 - this is actually not true as of vllm's current release: https://blog.vllm.ai/2024/09/05/perf-update.html

We also have more performance optimizations that will come out in v0.6.0:
- https://github.com/vllm-project/vllm/pull/8157

**Comment by stas00** [comment]
> @stas00 is this a new issue on v0.6.0?

no, same with older versions - e.g. 0.5.5

>  looks like someone has just fixed the issue:

Robert, I have already tried it to no avail.

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks @stas00 - @simon-mo can you copy in the folks who implemented the guided decoding?

**Comment by stas00** [comment]
@robertgshaw2-neuralmagic, do you know why `outlines` cache doesn't get used and instead it rebuilds the same. e.g. here is one of the repro scripts I created:

```
import time

from transformers import AutoTokenizer
from vllm import LLM, SamplingParams

model_name_or_path = "TinyLlama/TinyLlama-1.1B-Chat-v0.6"

schema = '{ "type": "object", "properties": { "name": { "type": "string", "minLength": 2,  "maxLength": 5 }, "age": { "type": "integer"} }, "required": ["name", "age"] }'

model = LLM(
    model=model_name_or_path,
    tokenizer=model_name_or_path,
    tokenizer_mode="auto",
    tensor_parallel_size=1,
    trust_remote_code=True,
    dtype="bfloat16",
    gpu_memory_utilization=0.8,
    guided_decoding_backend="outlines",
)

prompt = "Give an example of a person's profile that fits this JSON schema: {schema}"

sampling_params = SamplingParams(
    temperature=0.0,
    seed=42,
    max_tokens=2048,
)
kwargs = dict(
    sampling_params=sampling_params,
    guided_options_request=dict(
        guided_json=schema,
    ),
)

print("Message sent to vllm:")
print(prompt)

# warmup
response = model.generate(prompt, **kwargs)

response = ''
start_time = time.time()
for i in range(5):
    output = model.generate(prompt, **kwargs)
    response += output[0].outputs[0].text
end_time = time.time()

print(response)

tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)

response_tokens = tokenizer.encode(response)
response_tokens_len = len(response_tokens)
#print(response_tokens)

duration = end_time - start_time
print(f"decode tokens: {response_tokens_len} tokens")
print(f"decode tokens: {duration:0.2f}secs")
print(f"decode tokens: {response_tokens_len/duration:0.2f} tok/sec")
```
w/o warmup we get about 6 tok/sec due to FSM setup overhead. if we warm up it's 166 tok/sec.

re-running it the 2nd time w/o warm up repeats the same recompilation - shouldn't that be pulled from the cache and re-used, rather than recompiling it?

**Comment by Lap1n** [comment]
@stas00 I also got the same issue, seems like an issue with the key's format that Outlines uses to store cached data. I opened a [PR to fix on Outlines](https://github.com/outlines-dev/outlines/pull/1129)' repo.

**Comment by stas00** [comment]
so I'm waiting for my colleague to give me a repro case so that I could narrow it down for him.

Meanwhile, why is this pure `outlines` script so slow?

```
import outlines
import time
from transformers import AutoTokenizer
import json

model_name_or_path = "TinyLlama/TinyLlama-1.1B-Chat-v0.6"

schema = '{ "type": "object", "properties": { "name": { "type": "string", "minLength": 2,  "maxLength": 5 }, "age": { "type": "integer"} }, "required": ["name", "age"] }'

model = outlines.models.transformers(model_name_or_path)
generator = outlines.generate.json(model, schema)

prompt = "Give an example of a person's profile that fits this JSON schema: {schema}"

# warmup
response = generator(prompt)

# real
response = ''
start_time = time.time()
for i in range(5):
    response += json.dumps(generator(prompt))
end_time = time.time()
print(response)

tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)

response_tokens = tokenizer.encode(response)
response_tokens_len = len(response_tokens)
print(response_tokens)

duration = end_time - start_time
print(f"decode had {response_tokens_len} tokens")
print(f"decode latency: {duration:0.2f}secs")
print(f"decode throughput: {response_tokens_len/duration:0.2f} tok/sec")
```

I wrote it first to check that standalone `outlines` was fast and use it as a base-line before I did the exact same generation with vllm+outlines as I shared here:
https://github.com/vllm-project/vllm/issues/8313#issuecomment-2341897511

but actually I discovered that it's very slow!

On A100:
```
$ python outlines-direct.py
{"name": "John", "age": 29}{"name": "John", "age": 30}{"name": "John", "age": 50}{"name": "Dan", "age": 20}{"name": "John", "age": 30}
decode tokens: 66 tokens
decode tokens: 14.60secs
decode tokens: 4.52 tok/sec

$ python outlines-vllm.py 
{"name":"John","age":25}{"name":"John","age":25}{"name":"John","age":25}{"name":"John","age":25}{"name":"John","age":25}
decode tokens: 51 tokens
decode tokens: 0.27secs
decode tokens: 186.48 tok/sec
```

that's a 40x difference! Perhaps this is what we are hitting, but this is in reverse - vllm+outlines is 40x faster than outlines on its own.

**Comment by stas00** [comment]
the other odd thing I noticed is that vllm strips spaces:
```
# outlines:
{"name": "John", "age": 29}{"name": "John", "age": 30}

# vllm:
{"name":"John","age":25}{"name":"John","age":25}
```
I guess that saves a few tokens from needing to be generated w/o compromising the output structure - nice!

**Comment by stas00** [comment]
ok, I see that vllm isn't using `outlines.generate.json` generator but uses the regex engine. which has multiple issues:

1. why does it recompile the regex on every request? 

https://github.com/vllm-project/vllm/blob/b1f3e189586dce42bb3dcda20169a9308c9a25fa/vllm/model_executor/guided_decoding/outlines_logits_processors.py#L142

this should be cached!

2. this leads to vllm+outline failing to complete json last characters when max_tokens isn't too long. whereas `outlines.generate.json` seems to always return the correct json.

The problem here is that if the generator runs out of max_tokens and the json structure isn't closed - i.e. missing say `}` vllm will just return a broken json.

todo: add a repro script here.

**Comment by stas00** [comment]
The summary of things so far:

1. one problem proved to be a repeated recomputation of FSM in outlines because the cache was being ignored. @Lap1n has proposed a PR to fix it here https://github.com/dottxt-ai/outlines/pull/1129

2. the other issue is re-compiling of the regex on each request - I proposed to add it to this other PR by @Lap1n: https://github.com/vllm-project/vllm/pull/8308#issuecomment-2342517212

3. the truncated JSON is being actively discussed in this Issue: https://github.com/vllm-project/vllm/issues/8350

For the bigger picture that started the current main issue in the OP I'm blocked by my colleague to give me a larger repro test case, so that I could verify it and if true reduce it to a repro case, so please bear with me until they come through.

**Comment by Quang-elec44** [comment]
@

> > @stas00 In my experience, guided generation is always slower than normal. I recommend you try `sglang` instead. Sglang achieves better throughput than vLLM, but the guided generation is still slower.
> 
> Hey @Quang-elec44 - this is actually not true as of vllm's current release: https://blog.vllm.ai/2024/09/05/perf-update.html
> 
> We also have more performance optimizations that will come out in v0.6.0:
> 
> * [[Core][Bugfix][Perf] Introduce `MQLLMEngine` to avoid `asyncio` OH #8157](https://github.com/vllm-project/vllm/pull/8157)

Thanks for your information. I know that `vllm=0.06.0` made strong progress compared to `sglang=0.3.0`. However, in my experiments, `sglang` still achieves better throughput (my GPUs are A30). I may spend time testing the benchmark on my machine to have a closer look.

**Comment by Muhtasham** [comment]
Also noticed for very large JSON schemas using `guided_json` vLLM even in debug mode just returns 500 without any log

**Comment by aabbccddwasd** [comment]
any progress？

**Comment by dbuades** [comment]
We are experimenting the same issues on our side, guided generation is a great feature but it is very slow in offline mode.

**Comment by DanielRoeder1** [comment]
Is there any update on this?

I am running the samples from the docs for the [guided_grammar](https://github.com/vllm-project/vllm/blob/4fb8e329fd6f51d576bcf4b7e8907e0d83c4b5cf/examples/openai_chat_completion_structured_outputs.py#L70) using the newest docker container with a "meta-llama/Llama-3.2-1B-Instruct". 

Without guided grammar: 2.2 second response with 196 tokens
With guided grammar: Stopped generation after 20 minutes

**Comment by VirtualRoyalty** [comment]
Hi,
is this issue still relevant?
